presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -659,7 +659,8 @@ private void addTaskDynamicFilters(TaskId taskId, Map<DynamicFilterId, Domain> n
                 // Narrowing down of task dynamic filter is not supported.
                 // Currently, task dynamic filters are derived from join and semi-join,
                 // which produce just a single version of dynamic filter.
-                checkState(taskDomains.put(taskId, domain) == null, "Task dynamic filter is set twice");
+                Domain previousDomain = taskDomains.put(taskId, domain);
+                checkState(previousDomain == null || domain.equals(previousDomain), "Different task domains were set");
```

Comment:
Change looks ok to me.
One doubt, why don't we just do `taskDomains.putIfAbsent` and avoid this status check ?

Reply:
> One doubt, why don't we just do taskDomains.putIfAbsent and avoid this status check ?

That would mean invariant that there is one version of DF is broken somewhere. We want to enforce that invariant.